### PR TITLE
feat: add FormData, text, and none body types to HTTP request workflow

### LINKED ISSUE #1256: Support more body types in HTTP Request action
## Scope & Context

I started to do some work on this branch: https://github.com/twentyhq/twenty/compare/main...bd-json-fields

## Current behavior

Only JSON bodies can be sent.

## Expected behavior

We might want to send other types of request bodies, like form data. It would be great to support out of the box the following types:

- application/json
- multipart/form-data
- application/x-www-form-urlencoded
- text/plain

It corresponds to the options available in Postman.

Image

The form could look like this:

https://github.com/user-attachments/assets/a751c810-22f2-4ba2-afe0-eac52e2f8934

### Ensure retrocompatibility

All bodies were JSON previously, we must ensure they keep working while we add more options; it might involve writing a migration command

### `Content-Type` header management

Postman sets some headers by default but hides them. When selecting a specific body type, a default `Content-Type` will be set, such as `application/json`. Postman allows users to override this default `Content-Type`. We should determine how we want to handle this.

One great thing about axios is that once you've set the `Content-Type` of your request, it will serialize the body correctly.

### PR DESCRIPTION
**Pull request by omarNaifer12**
resolve twentyhq/core-team-issues#1256

https://github.com/user-attachments/assets/d53faccf-dfa0-4425-a1a0-02f71f97fa19

### CODE REVIEW

**Comment by charlesBochet** [comment]
@omarNaifer12 thanks for opening a PR

All CIs are red, could you take a look?

File: packages/twenty-front/src/modules/workflow/validation-schemas/workflowSchema.ts [hunk 1]
```diff
@@ -127,11 +127,19 @@ export const workflowHttpRequestActionSettingsSchema =
...
+                path: z.string(),
+                filename: z.string(),
+              }),
+            ),
           ]),
         )
         .or(z.string())
         .optional(),
+      bodyType: z
+        .enum(['keyValue', 'rawJson', 'FormData', 'Text', 'None'])
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/validation-schemas/workflowSchema.ts:141]: why adding a bodyType here? I don't think this is requested in the ticket? I thought we wanted to rely on the headers input

File: packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/components/BodyInput.tsx [hunk 2]
```diff
@@ -27,33 +29,51 @@ const StyledContainer = styled.div`
...
+  bodyType,
 }: BodyInputProps) => {
   const defaultValueParsed = isString(defaultValue)
     ? (parseJson<JsonValue>(defaultValue) ?? {})
     : defaultValue;
 
-  const [isRawJson, setIsRawJson] = useState(
-    shouldDisplayRawJsonByDefault(defaultValue),
+  const [bodyTypeValue, setBodyTypeValue] = useState<BodyType>(
+    bodyType || 'keyValue',
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/components/BodyInput.tsx:58]: this could be parsed from headers as well

File: packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/components/MultiPartFormDataInput.tsx [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+  readonly?: boolean;
+  keyPlaceholder?: string;
+};
+
+const inputTypeOptions: SelectOption<'text' | 'file'>[] = [
+  { value: 'text', label: 'Text' },
+  { value: 'file', label: 'File' },
+];
+
+export const MultiPartFormDataInput = ({
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/components/MultiPartFormDataInput.tsx:126]: do we actually need a new component ? Checking the ticket, it feels like we are using the same kind of form as headers

File: packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/graphqlMutation/httpRequestMutation.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { gql } from '@apollo/client';
+
+export const HTTP_REQUEST_MUTATION = gql`
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/workflow-steps/workflow-actions/http-request-action/graphqlMutation/httpRequestMutation.ts:3]: we already have a mutation to test http request, why do we need a new one?

File: packages/twenty-server/src/engine/metadata-modules/workflow-version-step-http-request-test/workflow-version-step-http-request-test.resolver.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { AuthWorkspace } from 'src/engine/decorators/auth/auth-workspace.decorator';
+import { WorkspaceAuthGuard } from 'src/engine/guards/workspace-auth.guard';
+import { TestHttpResponseDto } from 'src/engine/metadata-modules/workflow-version-step-http-request-test/dtos/test-step-http-response.dto';
+import { WorkflowHttpRequestActionInputDto } from 'src/engine/metadata-modules/workflow-version-step-http-request-test/dtos/workflow-http-request-action-input.dto';
+
+import { WorkflowVersionStepHttpRequestTestService } from './workflow-version-step-http-request-test.service';
+
+@Resolver()
+@UseGuards(WorkspaceAuthGuard)
+export class WorkflowVersionStepHttpRequestTestResolver {
```
Comment (thomtrp) [inline packages/twenty-server/src/engine/metadata-modules/workflow-version-step-http-request-test/workflow-version-step-http-request-test.resolver.ts:14]: Why do we need this resolver? Testing http request could remain on frontend side

**Review by thomtrp — COMMENTED** [review]
@omarNaifer12 thank you for your PR. I feel this code is a bit too complicated for what we want to achieve. 

Here are general feedbacks:
- keep simple code, only frontend. If we need backend for some content, let's not support those for now
- avoid building new big frontend components. I feel like the design is quite similar to what is already existing
- avoid adding code everywhere just to support this test http request use case. Step should be generic and except for a few of them, there should not be edge cases everywhere. Otherwise it will not be maintainable

**Comment by omarNaifer12** [comment]
Thanks @thomtrp for the review  I understand your point. Should I use headers instead of bodyType, reuse existing components instead of creating new ones, and therefore avoid creating a new resolver?

At first, when I was thinking about testing the API with FormData, I wasn’t sure whether to handle it on the backend or the frontend. My concern was: if I handle it on the frontend, where would I get the files data?I wasn’t sure,but I thought it might be unnecessary to fetch all the files together with the other data shown to the user when they open the workflow version to see the steps, since the files are only needed when testing, and they could be large.

Another option I considered was fetching the files only when the user opens the HTTP request step, but that also seemed unnecessary.

The last solution I thought of was to fetch all files from the backend when the user clicks "Test" with multipart FormData, then use those files directly in the fetch request. But in this case, the backend would still need a resolver to provide all the files with their exact keys, because we wouldn’t know which key belongs to which file if we just fetch them without metadata. This means I might need to create a new resolver  or, could I instead use the existing API controller (Get('*/:filename')), fetch all files with Promise.all on the frontend, and then build an array of objects to preserve the FormData keys along with the files?

Could you suggest which approach is best in our situation?

Also, can I go ahead and fix all of this, clean up the code, and then create a new PR?

One last thing: could you please also check the http-tool.ts file and see how I solved it with FormData? Is that approach correct, or should it also be reworked? Right now, I fetch all files with async.all, build a FormData object with the keys, buffer, and filename, and then return the FormData ready to use in Axios.

**Comment by thomtrp** [comment]
> Should I use headers instead of bodyType, reuse existing components instead of creating new ones, and therefore avoid creating a new resolver?

Yes, body type should be directly depending on headers. See video in the ticket.
About new components, you can create one if you feel this is cleaner but it should remain really simple.

> where would I get the files data?

Do not care about files data yet. Let's implement a first form data version that do not support files. Not sure we want to support it there anyway. We need to discuss it with @Bonapara once he's back from PTO. And if we do, we will provide a more accurate strategy

> Also, can I go ahead and fix all of this, clean up the code, and then create a new PR?
 
 Of course ! Once code is cleaned feel free to re-open 👍

**Comment by omarNaifer12** [comment]
@thomtrp I did some investigation, but I’m not sure how to rely only on headers to know which body type option the user selected. How can I differentiate between a Content-Type header typed by the user and the one I saved in the database after they chose a body type from the input list? And if I don’t save the Content-Type header when the user changes the body type option (without typing a header manually), what should I save in the database to track which body type they’re using? Could you suggest any approach?
